src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java

```diff
@@ -1,25 +1,110 @@
 package com.es.phoneshop.model.product;
 
+import java.math.BigDecimal;
+import java.util.ArrayList;
+import java.util.Currency;
 import java.util.List;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.stream.Collectors;
 
 public class ArrayListProductDao implements ProductDao {
+    private long currentId;
+    private List<Product> products;
+    private final Lock productsLock = new ReentrantLock();
+
+
+    public ArrayListProductDao() {
+        this.products = new ArrayList<>();
+        saveSampleProducts();
+    }
+
     @Override
-    public Product getProduct(Long id) {
-        throw new RuntimeException("Not implemented");
+    public Product getProduct(Long id) throws ProductNotFoundException {
+        productsLock.lock();
+        try {
+            return products.stream()
+                    .filter(product -> id.equals(product.getId()))
+                    .findAny()
+                    .orElseThrow(ProductNotFoundException::new);
+        }
+        finally {
+            productsLock.unlock();
+        }
     }
 
     @Override
     public List<Product> findProducts() {
-        throw new RuntimeException("Not implemented");
+        productsLock.lock();
+        try {
+            return products.stream()
+                    .filter(product -> product.getPrice() != null)
+                    .filter(product -> product.getStock() > 0)
+                    .collect(Collectors.toList());
+        }
+        finally {
+            productsLock.unlock();
+        }
     }
 
     @Override
-    public void save(Product product) {
-        throw new RuntimeException("Not implemented");
+    public void save(Product product) throws ProductNotFoundException {
+        productsLock.lock();
+        try {
+            if (product.getId() != null) {
+                products.stream()
+                        .filter(product1 -> product1.getId().equals(product.getId()))
+                        .findAny()
+                        .map(product1 -> products.set(products.indexOf(product1), product))
+                        .orElseThrow(ProductNotFoundException::new);
+            } else {
+                product.setId(++currentId);
+                products.add(product);
+            }
+        }
+        finally {
+            productsLock.unlock();
+        }
     }
 
     @Override
-    public void delete(Long id) {
-        throw new RuntimeException("Not implemented");
+    public void delete(Long id) throws ProductNotFoundException {
+        productsLock.lock();
+        try {
+            products.stream()
+                    .filter(product -> id.equals(product.getId()))
+                    .findAny()
+                    .map(product -> products.remove(product))
+                    .orElseThrow(ProductNotFoundException::new);
+            products.stream()
```

Comment:
you are deleting one product
why do you need to rewrite all other products? It's unnecessary and dangerous 
ID of objects are constant otherwise we may fall with hundreds of unexpected errors